runners/spark/3/src/main/java/org/apache/beam/runners/spark/structuredstreaming/translation/batch/DatasetSourceBatch.java

```diff
@@ -0,0 +1,240 @@
+import java.util.Map;
+import java.util.Set;
+import org.apache.beam.runners.core.construction.SerializablePipelineOptions;
+import org.apache.beam.runners.core.serialization.Base64Serializer;
+import org.apache.beam.runners.spark.structuredstreaming.translation.helpers.RowHelpers;
+import org.apache.beam.runners.spark.structuredstreaming.translation.helpers.SchemaHelpers;
+import org.apache.beam.sdk.io.BoundedSource;
+import org.apache.beam.sdk.options.PipelineOptions;
+import org.apache.beam.sdk.util.WindowedValue;
+import org.apache.parquet.Strings;
+import org.apache.spark.sql.catalyst.InternalRow;
+import org.apache.spark.sql.connector.catalog.SupportsRead;
+import org.apache.spark.sql.connector.catalog.Table;
+import org.apache.spark.sql.connector.catalog.TableCapability;
+import org.apache.spark.sql.connector.catalog.TableProvider;
+import org.apache.spark.sql.connector.expressions.Transform;
+import org.apache.spark.sql.connector.read.Batch;
+import org.apache.spark.sql.connector.read.InputPartition;
+import org.apache.spark.sql.connector.read.PartitionReader;
+import org.apache.spark.sql.connector.read.PartitionReaderFactory;
+import org.apache.spark.sql.connector.read.Scan;
+import org.apache.spark.sql.connector.read.ScanBuilder;
+import org.apache.spark.sql.types.StructType;
+import org.apache.spark.sql.util.CaseInsensitiveStringMap;
+
+/**
+ * Empty impl needed for compilation. Spark DataSourceV2 API was removed in Spark3. Need to code a
+ * Beam source using new spark 3 API.
+ */
+public class DatasetSourceBatch implements TableProvider {
+
+  private static final StructType BINARY_SCHEMA = SchemaHelpers.binarySchema();
+
+  public DatasetSourceBatch() {}
+
+  @Override
+  public StructType inferSchema(CaseInsensitiveStringMap options) {
+    return BINARY_SCHEMA;
+  }
+
+  @Override
+  public boolean supportsExternalMetadata() {
+    return true;
+  }
+
+  @Override
+  public Table getTable(
+      StructType schema, Transform[] partitioning, Map<String, String> properties) {
+    return new DatasetSourceBatchTable();
+  }
+
+  private static class DatasetSourceBatchTable implements SupportsRead {
+
+    @Override
+    public ScanBuilder newScanBuilder(CaseInsensitiveStringMap options) {
+      return new ScanBuilder() {
+
+        @Override
+        public Scan build() {
+          return new Scan() { // scan for Batch reading
+
+            @Override
+            public StructType readSchema() {
+              return BINARY_SCHEMA;
+            }
+
+            @Override
+            public Batch toBatch() {
+              return new BeamBatch<>(options);
+            }
+          };
+        }
+      };
+    }
+
+    @Override
+    public String name() {
+      return "BeamSource";
+    }
+
+    @Override
+    public StructType schema() {
+      return BINARY_SCHEMA;
+    }
+
+    @Override
+    public Set<TableCapability> capabilities() {
+      final HashSet<TableCapability> capabilities = new HashSet<>();
```

Comment:
Nit: can we make this an ImmutableSet and statically initialize it?

Reply:
I tried to stick to java libs, but sure, we can use guava as it is already used everywhere in beam